internal/liveshare/test/server.go

```diff
@@ -145,73 +155,145 @@ func makeConnection(server *Server) http.HandlerFunc {
 				return
 			}
 		}
 		c, err := upgrader.Upgrade(w, req, nil)
 		if err != nil {
-			server.errCh <- fmt.Errorf("error upgrading connection: %w", err)
+			sendError(server.errCh, fmt.Errorf("error upgrading connection: %w", err))
 			return
 		}
-		defer c.Close()
+		defer func() {
+			if err := c.Close(); err != nil {
+				sendError(server.errCh, err)
+			}
+		}()
 
 		socketConn := newSocketConn(c)
 		_, chans, reqs, err := ssh.NewServerConn(socketConn, server.sshConfig)
 		if err != nil {
-			server.errCh <- fmt.Errorf("error creating new ssh conn: %w", err)
+			sendError(server.errCh, fmt.Errorf("error creating new ssh conn: %w", err))
 			return
 		}
 		go ssh.DiscardRequests(reqs)
 
-		for newChannel := range chans {
-			ch, reqs, err := newChannel.Accept()
+		if err := handleChannels(ctx, server, chans); err != nil {
+			sendError(server.errCh, err)
+		}
+	}
+}
+
+func sendError(errc chan<- error, err error) {
+	select {
+	case errc <- err:
+	default:
+	}
+}
+
+func awaitError(ctx context.Context, errc <-chan error) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errc:
+		return err
+	}
+}
+
+func handleChannels(ctx context.Context, server *Server, sshChannels <-chan ssh.NewChannel) error {
+	errc := make(chan error, 1)
+	go func() {
+		for sshCh := range sshChannels {
+			ch, reqs, err := sshCh.Accept()
 			if err != nil {
-				server.errCh <- fmt.Errorf("error accepting new channel: %w", err)
+				sendError(errc, fmt.Errorf("failed to accept channel: %w", err))
 				return
 			}
-			go handleNewRequests(ctx, server, ch, reqs)
-			go handleNewChannel(server, ch)
+
+			go func() {
+				if err := handleRequests(ctx, server, ch, reqs); err != nil {
+					sendError(errc, fmt.Errorf("failed to handle requests: %w", err))
+				}
+			}()
+
+			handleChannel(server, ch)
 		}
-	}
+	}()
+	return awaitError(ctx, errc)
 }
 
-func handleNewRequests(ctx context.Context, server *Server, channel ssh.Channel, reqs <-chan *ssh.Request) {
-	for req := range reqs {
-		if req.WantReply {
-			if err := req.Reply(true, nil); err != nil {
-				server.errCh <- fmt.Errorf("error replying to channel request: %w", err)
+func handleRequests(ctx context.Context, server *Server, channel ssh.Channel, reqs <-chan *ssh.Request) error {
+	errc := make(chan error, 1)
+	go func() {
+		for req := range reqs {
+			if req.WantReply {
+				if err := req.Reply(true, nil); err != nil {
+					sendError(errc, fmt.Errorf("error replying to channel request: %w", err))
+					return
+				}
+			}
+
+			if strings.HasPrefix(req.Type, "stream-transport") {
+				go func() {
+					if err := forwardStream(ctx, server, req.Type, channel); err != nil {
+						sendError(errc, fmt.Errorf("failed to forward stream: %w", err))
+					}
+				}()
 			}
 		}
-		if strings.HasPrefix(req.Type, "stream-transport") {
-			forwardStream(ctx, server, req.Type, channel)
-		}
-	}
+	}()
+
+	return awaitError(ctx, errc)
```

Comment:
Perhaps I'm missing something, but why would we start a goroutine and then immediately wait for it to finish to collect its error object? Why not do it synchronously?